// myst-directives/dist/iframe.js
import { normalizeLabel } from 'myst-common';
export const iframeDirective = {
    name: 'iframe',
    arg: {
        type: String,
        doc: 'The URL source (`src`) of the HTML iframe element.',
        required: true,
    },
    options: {
        label: {
            type: String,
            alias: ['name'],
        },
        class: {
            type: String,
            // class_option: list of strings?
        },
        width: {
            type: String,
            doc: 'The iframe width, in CSS units, for example `50%` or `300px`.',
        },
        align: {
            type: String,
            doc: 'The alignment of the iframe in the page. Choose one of `left`, `center` or `right`',
        },
    },
    body: { type: 'myst', doc: 'If provided, this will be the iframe caption.' },
    run(data) {
        var _a, _b, _c, _d, _e;
        const { label, identifier } = normalizeLabel((_a = data.options) === null || _a === void 0 ? void 0 : _a.label) || {};
        const iframe = {
            type: 'iframe',
            src: data.arg,
            width: (_b = data.options) === null || _b === void 0 ? void 0 : _b.width,
            align: (_c = data.options) === null || _c === void 0 ? void 0 : _c.align,
        };
        if (!data.body) {
            iframe.label = label;
            iframe.identifier = identifier;
            iframe.class = (_d = data.options) === null || _d === void 0 ? void 0 : _d.class;
            return [iframe];
        }
        const container = {
            type: 'container',
            kind: 'figure',
            identifier,
            label,
            class: (_e = data.options) === null || _e === void 0 ? void 0 : _e.class,
            children: [iframe, { type: 'caption', children: data.body }],
        };
        return [container];
    },
};
