// myst-roles/dist/si.js
export const siRole = {
    name: 'si',
    body: {
        type: String,
        required: true,
    },
    run(data) {
        const value = data.body;
        const match = value.match(/([0-9]+)\s?<([\\a-zA-Z\s]+)>/);
        if (!match) {
            return [{ type: 'si', error: true, value }];
        }
        const [, number, commands] = match;
        const parsed = [...commands.matchAll(/\\([a-zA-Z]+)/g)];
        const units = parsed.filter((c) => !!c).map(([, c]) => c);
        const translated = units.map((c) => { var _a; return (_a = UNITS[c]) !== null && _a !== void 0 ? _a : c; });
        const si = {
            type: 'si',
            number,
            unit: translated.join(''),
            units,
            alt: units.join(' '),
            value: `${number} ${translated.join('')}`,
        };
        return [si];
    },
};
const UNITS = {
    ampere: 'A',
    candela: 'cd',
    kelvin: 'K',
    kilogram: 'kg',
    metre: 'm',
    meter: 'm',
    mole: 'mol',
    second: 's',
    becquerel: 'Bq',
    degreeCelsius: '°C',
    coulomb: 'C',
    farad: 'F',
    gray: 'Gy',
    hertz: 'Hz',
    henry: 'H',
    joule: 'J',
    lumen: 'lm',
    katal: 'kat',
    lux: 'lx',
    newton: 'N',
    ohm: 'Ω',
    pascal: 'Pa',
    radian: 'rad',
    siemens: 'S',
    sievert: 'Sv',
    steradian: 'sr',
    tesla: 'T',
    volt: 'V',
    watt: 'W',
    weber: 'Wb',
    astronomicalunit: 'au',
    bel: 'B',
    dalton: 'Da',
    day: 'd',
    decibel: 'dB',
    degree: '°',
    electronvolt: 'eV',
    hectare: 'ha',
    hour: 'h',
    litre: 'L',
    liter: 'L',
    arcminute: '′',
    minute: 'min',
    arcsecond: '″',
    neper: 'Np',
    tonne: 't',
    // SI prefixes
    yocto: 'y',
    zepto: 'z',
    atto: 'a',
    femto: 'f',
    pico: 'p',
    nano: 'n',
    micro: 'µ',
    milli: 'm',
    centi: 'c',
    deci: 'd',
    deca: 'da',
    hecto: 'h',
    kilo: 'k',
    mega: 'M',
    giga: 'G',
    tera: 'T',
    peta: 'P',
    exa: 'E',
    zetta: 'Z',
    yotta: 'Y',
    // Special
    angstrom: 'Å',
};
