// myst-common/dist/ruleids.js
export var RuleId;
(function (RuleId) {
    // Frontmatter rules
    RuleId["validConfigStructure"] = "valid-config-structure";
    RuleId["siteConfigExists"] = "site-config-exists";
    RuleId["projectConfigExists"] = "project-config-exists";
    RuleId["validSiteConfig"] = "valid-site-config";
    RuleId["validProjectConfig"] = "valid-project-config";
    RuleId["configHasNoDeprecatedFields"] = "config-has-no-deprecated-fields";
    RuleId["frontmatterIsYaml"] = "frontmatter-is-yaml";
    RuleId["validPageFrontmatter"] = "valid-page-frontmatter";
    RuleId["validFrontmatterExportList"] = "valid-frontmatter-export-list";
    // Export rules
    RuleId["docxRenders"] = "docx-renders";
    RuleId["jatsRenders"] = "jats-renders";
    RuleId["mdRenders"] = "md-renders";
    RuleId["mecaIncludesJats"] = "meca-includes-jats";
    RuleId["mecaExportsBuilt"] = "meca-exports-built";
    RuleId["mecaFilesCopied"] = "meca-files-copied";
    RuleId["pdfBuildCommandsAvailable"] = "pdf-build-commands-available";
    RuleId["pdfBuildsWithoutErrors"] = "pdf-builds-without-errors";
    RuleId["pdfBuilds"] = "pdf-builds";
    RuleId["texRenders"] = "tex-renders";
    RuleId["exportExtensionCorrect"] = "export-extension-correct";
    RuleId["exportArticleExists"] = "export-article-exists";
    // Parse rules
    RuleId["texParses"] = "tex-parses";
    RuleId["jatsParses"] = "jats-parses";
    RuleId["mystFileLoads"] = "myst-file-loads";
    RuleId["selectedFileIsProcessed"] = "selected-file-is-processed";
    // Directive and role rules
    RuleId["directiveRegistered"] = "directive-registered";
    RuleId["directiveKnown"] = "directive-known";
    RuleId["directiveArgumentCorrect"] = "directive-argument-correct";
    RuleId["directiveOptionsCorrect"] = "directive-options-correct";
    RuleId["directiveBodyCorrect"] = "directive-body-correct";
    RuleId["roleRegistered"] = "role-registered";
    RuleId["roleKnown"] = "role-known";
    RuleId["roleBodyCorrect"] = "role-body-correct";
    // Project structure rules
    RuleId["tocContentsExist"] = "toc-contents-exist";
    RuleId["validTocStructure"] = "valid-toc-structure";
    RuleId["validToc"] = "valid-toc";
    RuleId["tocWritten"] = "toc-written";
    // Image rules
    RuleId["imageDownloads"] = "image-downloads";
    RuleId["imageExists"] = "image-exists";
    RuleId["imageFormatConverts"] = "image-format-converts";
    RuleId["imageCopied"] = "image-copied";
    RuleId["imageFormatOptimizes"] = "image-format-optimizes";
    // Math rules
    RuleId["mathLabelLifted"] = "math-label-lifted";
    RuleId["mathEquationEnvRemoved"] = "math-equation-env-removed";
    RuleId["mathEqnarrayReplaced"] = "math-eqnarray-replaced";
    RuleId["mathAlignmentAdjusted"] = "math-alignment-adjusted";
    RuleId["mathRenders"] = "math-renders";
    // Reference rules
    RuleId["referenceTemplateFills"] = "reference-template-fills";
    RuleId["identifierIsUnique"] = "identifier-is-unique";
    RuleId["referenceTargetResolves"] = "reference-target-resolves";
    RuleId["referenceSyntaxValid"] = "reference-syntax-valid";
    RuleId["referenceTargetExplicit"] = "reference-target-explicit";
    RuleId["footnoteReferencesDefinition"] = "footnote-references-definition";
    RuleId["intersphinxReferencesResolve"] = "intersphinx-references-resolve";
    // Link rules
    RuleId["mystLinkValid"] = "myst-link-valid";
    RuleId["rridLinkValid"] = "rrid-link-valid";
    RuleId["wikipediaLinkValid"] = "wikipedia-link-valid";
    RuleId["doiLinkValid"] = "doi-link-valid";
    RuleId["linkResolves"] = "link-resolves";
    // Notebook rules
    RuleId["notebookAttachmentsResolve"] = "notebook-attachments-resolve";
    RuleId["notebookOutputCopied"] = "notebook-output-copied";
    // Content rules
    RuleId["mdastSnippetImports"] = "mdast-snippet-imports";
    RuleId["includeContentFilters"] = "include-content-filters";
    RuleId["includeContentLoads"] = "include-content-loads";
    RuleId["gatedNodesJoin"] = "gated-nodes-join";
    RuleId["glossaryUsesDefinitionList"] = "glossary-uses-definition-list";
    RuleId["blockMetadataLoads"] = "block-metadata-loads";
    // Citation rules
    RuleId["citationIsUnique"] = "citation-is-unique";
    RuleId["bibFileExists"] = "bib-file-exists";
    RuleId["citationRenders"] = "citation-renders";
    // Code rules
    RuleId["codeMetadataLifted"] = "code-metadata-lifted";
    RuleId["codeMetatagsValid"] = "code-metatags-valid";
    RuleId["codeLangDefined"] = "code-lang-defined";
    RuleId["codeMetadataLoads"] = "code-metadata-loads";
    RuleId["inlineCodeMalformed"] = "inline-code-malformed";
    RuleId["inlineExpressionRenders"] = "inline-expression-renders";
    // Static file rules
    RuleId["staticFileCopied"] = "static-file-copied";
    RuleId["exportFileCopied"] = "export-file-copied";
    RuleId["sourceFileCopied"] = "source-file-copied";
    RuleId["templateFileCopied"] = "template-file-copied";
    RuleId["staticActionFileCopied"] = "static-action-file-copied";
    // Plugins
    RuleId["pluginLoads"] = "plugin-loads";
    // Container rules
    RuleId["containerChildrenValid"] = "contianer-children-valid";
})(RuleId || (RuleId = {}));
